### PR TITLE
Sticky sessions

### PR DESCRIPTION
**Pull request by jmelis**
The goal of this PR is to enable session reusage in the graphql queries.

We were using the [GraphQLClient](https://pypi.org/project/graphqlclient/), but the codebase seems abandoned, and it doesn't support reusing sessions, so this PR has replaced the external `GraphQLClient` with an internal `QontractServerClient`.

It's important to note that the session will be regenerated every time we `init` again, which is good, because the [run-integration.py](https://github.com/jmelis/qontract-reconcile/blob/sticky-sessions/dockerfiles/hack/run-integration.py#L69) script will reinitialize every time. This has been tested (although manually, no automated tests for this).

### CODE REVIEW

File: reconcile/utils/gql.py [hunk 1]
```diff
@@ -91,20 +83,13 @@ def __init__(self, url, token=None, int_name=None, validate_schemas=False):
             if not self._valid_schemas:
                 raise GqlApiIntegrationNotFound(int_name)
 
-    @retry(exceptions=GqlApiError, max_attempts=5)
```
Comment (jmelis) [inline reconcile/utils/gql.py:94]: Removing the `@retry` decorator. I feel we should let it crash every time the server is not reachable in order not to obscure the errors.

File: reconcile/utils/gql.py [hunk 2]
```diff
@@ -91,20 +83,13 @@ def __init__(self, url, token=None, int_name=None, validate_schemas=False):
...
     def query(self, query, variables=None, skip_validation=False):
         try:
-            # supress print on HTTP error
-            # https://github.com/prisma-labs/python-graphql-client
-            # /blob/master/graphqlclient/client.py#L32-L33
-            with open(os.devnull, 'w') as f, contextlib.redirect_stdout(f):
-                result_json = self.client.execute(query, variables)
+            result = self.client.query(query, variables)
```
Comment (jmelis) [inline reconcile/utils/gql.py:88]: No need to suppress print on HTTP error as the new library is not doing that.

File: reconcile/utils/qontract_server_client.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        self.use_sessions = use_sessions
+        self.sha = None
+
+        if sha_url:
+            self.set_graphqlsha_url()
+        else:
+            self.query_url = url
+
+        if self.use_sessions:
+            self.session = requests.Session()
```
Comment (jmelis) [inline reconcile/utils/qontract_server_client.py:23]: This is the main purpose of this PR.

File: reconcile/utils/qontract_server_client.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        :rtype: dict
+        """
+        data = {'query': query, 'variables': variables}
+
+        headers = {'Accept': 'application/json',
+                   'Content-Type': 'application/json'}
+        if self.token:
+            headers['Authorization'] = self.token
+
+        if self.use_sessions:
```
Comment (apahim) [inline reconcile/utils/qontract_server_client.py:77]: You don't really need `self.use_sessions` as a class attribute. You should control that with `self.session`, having it initialized as `None` or as `requests.Session()`. Here you'd check `if/else` `self.session is None`.

Comment (jmelis) [reply reconcile/utils/qontract_server_client.py]: fixed

File: reconcile/utils/gql.py (hunk 1 above)
Comment (jmelis) [reply reconcile/utils/gql.py:94]: @maorfr do you have any reasons for not doing this?

Comment (maorfr) [reply reconcile/utils/gql.py:94]: network issues?

if we're looking for server issues, let's instrument the server IMO

**Comment by apahim** [comment]
reviewing

File: reconcile/utils/gql.py [hunk 5]
```diff
@@ -162,51 +145,74 @@ def get_queried_schemas(self):
...
+        defaults to True
+    :type print_url: bool, optional
+    :param use_sessions: reuse session for all the requests performed.
+        defaults to True
+    :type use_sessions: bool, optional
+
+    """
+    global _gqlapi
+    _gqlapi = GqlApi(url, token, integration, validate_schemas,
+                     sha_url=sha_url, use_sessions=True)
```
Comment (apahim) [inline reconcile/utils/gql.py:203]: `use_sessions=use_sessions` ?

**Comment by maorfr** [comment]
big +1 for moving away from the current gql client

File: reconcile/utils/qontract_server_client.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
+from urllib.parse import urlparse
+
+import requests
+
+
+class QontractServerClient:
+    """REST API for Qontract Server """
+
+    def __init__(self, url, token=None,
+                 use_sessions=False,
```
Comment (apahim) [inline reconcile/utils/qontract_server_client.py:10]: the `use_sessions` argument doesn't tell much & is exposing internal implementation... I'd suggest calling it `sticky_session` instead.

File: reconcile/utils/qontract_server_client.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+        if sha_url:
+            self.set_graphqlsha_url()
+        else:
+            self.query_url = url
+
+    def set_graphqlsha_url(self):
+        """Fetches the latest sha and sets the query url to use it.
+        """
+        self.sha = self.get_sha()
```
Comment (apahim) [inline reconcile/utils/qontract_server_client.py:29]: It's a personal choice I guess, but I always find the class variables initialization outside of the `__init__` confusing. My suggestion is to return values from the helpers and to have all the initializations done on the `__init__`.

File: reconcile/utils/qontract_server_client.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        :type variables: dict, optional
+        :return: graphql response payload. keys: 'data' and '
+        :rtype: dict
+        """
+        data = {'query': query, 'variables': variables}
+
+        headers = {'Accept': 'application/json',
+                   'Content-Type': 'application/json'}
+        if self.token:
+            headers['Authorization'] = self.token
```
Comment (apahim) [inline reconcile/utils/qontract_server_client.py:76]: `headers` never changes... I'd move its initialization to `__init__`.

File: reconcile/utils/qontract_server_client.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+        r = post_method(self.query_url, json=data, headers=headers)
+        r.raise_for_status()
+        return r.json()
+
+    def _base_url_path(self, path):
+        return urlparse(self.base_url)._replace(path=path).geturl()
+
+    def _get(self, path):
+        headers = {'Authorization': self.token} if self.token else None
```
Comment (apahim) [inline reconcile/utils/qontract_server_client.py:91]: can we re-use the `headers` from `query()`?

**Review by apahim — APPROVED** [review]
This is awesome! Lgtm, with a few small suggestions.

File: reconcile/utils/gql.py (hunk 1 above)
Comment (jmelis) [reply reconcile/utils/gql.py:94]: would you be aginst trying this out? I'd like to. We can add the decorator if we see a lot of errors.

**Comment by apahim** [comment]
lgtm! ship it!

**Comment by jmelis** [comment]
thanks!

**Pull request merged**